Modules/Core/Common/src/itkObject.cxx

```diff
@@ -219,7 +208,7 @@ SubjectImplementation::InvokeEventRecursion(const EventObject &
   {
 
     // save observer
-    const Observer * o = *i;
+    const auto & o = *i;
```

Comment:
I'm not sure what the type is here? Reference to the unique pointer in the list?

The point of this patter is to save the pointer to the object, in case it was deleted/modified. Since the list still has the sole ownership I don't believe this works as intended any more. Set the prose above.

Reply:
Perhaps this should be maintained as a raw pointer, and the below "find" be modified to compare raw pointer of m_Observer items to this save raw pointer?

Reply:
Is `InvokeEventRecursion` exercised in current tests? If not, now would be a good time to cover it by test(s).

Reply:
There was a reasonable test added for this case:  f7307e26867076074a04b956633676964d5f7ffe It may be that valgrind will be required to detect the invalid memory access.

Reply:
Thanks @blowekamp The test that you provided, [CommandObserverObjectTest](https://github.com/InsightSoftwareConsortium/ITK/blob/2f2b29cbbe3fe91474933e0194db8d581d5b3247/Modules/Core/Common/test/itkCommandObserverObjectTest.cxx), is very helpful. I think it covers the internal helper function `InvokeEventRecursion`  well enough 👍 And it does pass successfully on this pull request  😄  

The pull request is not a design change. It just replaces existing `new` and `delete` commands by using `unique_ptr` instead. It does not change the behavior. IMhO, the use of  `unique_ptr` expresses more clearly (than when using raw pointers) that the list (`m_Observers`) has the sole ownership of the internal Observer objects.

Regarding the `std::find` call: https://github.com/InsightSoftwareConsortium/ITK/blob/2f2b29cbbe3fe91474933e0194db8d581d5b3247/Modules/Core/Common/src/itkObject.cxx#L228
This `std::find` call works just with a list of `unique_ptr`s  as with a list of raw pointers. (`std::unique_ptr` has an `operator==` which simply compares the internal raw pointers: https://en.cppreference.com/w/cpp/memory/unique_ptr/operator_cmp). 

You wrote about `const auto & o = *i`:
> I'm not sure what the type is here? Reference to the unique pointer in the list?

Indeed. Would you prefer it something like  `const ObserverPointerType & o = *i`? I still think this pull request ready to be merged right now. But if you prefer something like  `const ObserverPointerType & o = *i`, I can change that, of course. What do you think?

Reply:
> Perhaps this should be maintained as a raw pointer, and the below "find" be modified to compare raw pointer of m_Observer items to this save raw pointer?

Thanks @blowekamp This comment of yours inspired me to submit pull request #3629: "BUG: `Object::InvokeEvent` should not use a dangling Observer pointer"! Please check!

I think this PR (#3620) and PR #3629 can be processed in either order, but it's probably easiest now to first process PR #3629.